### PR TITLE
Let Term

### PR DESCRIPTION
**Pull request by HoshinoTented**
+ [x] `Let Term` and normalization
+ [x] `Nested` interface for nested structure
+ [x] `LetPrettier` for prettying let-like things

TODO:
+ [ ] rendering the type of the bind during rendering `LetTerm`. I may use Synthesizer, but where does the `TyckState` come from?

### CODE REVIEW

File: base/src/main/java/org/aya/concrete/Expr.java [hunk 1]
```diff
@@ -44,6 +44,43 @@
...
+     *
+     * @return null if the body is not {@link This}
+     * @implSpec {@code tryNested == null || tryNested == body}
+     */
+    @SuppressWarnings("unchecked")
+    default @Nullable This tryNested() {
+      var body = body();
+      var clazz = getClass();
+
+      return clazz.isInstance(body) ? (This) body : null;
```
Comment (ice1000) [inline base/src/main/java/org/aya/concrete/Expr.java:80]: clazz.cast maybe

**Review by ice1000 — COMMENTED** [review]
Very nice! I'll go make some minor edits and merge this.

Gj!

File: base/src/main/java/org/aya/concrete/Expr.java (hunk 1 above)
Comment (HoshinoTented) [reply base/src/main/java/org/aya/concrete/Expr.java:80]: `(This) clazz.cast(body)` ?

**Comment by HoshinoTented** [comment]
😢

**Comment by HoshinoTented** [comment]
What's wrong? Consider we want to `term[x |-> y]` with defEq `z |-> x`, the substituted term still contains `x` after normalization.

So I suspend this PR.

**Comment by HoshinoTented** [comment]
This feature may cost more than its benefit...

**Pull request closed without merging**